data/python.gram

```diff
@@ -171,59 +189,86 @@ class Parser(Parser):
         """Set the context (Load, Store, Del) of an ast node."""
         node.ctx = context
         return node
 
-    def ensure_real(self, number_str: str):
-        number = ast.literal_eval(number_str)
-        if type(number) is complex:
-            self.raise_syntax_error("real number required in complex literal")
-        return number
+    def ensure_real(self, number: ast.Constant):
+        value = ast.literal_eval(number.string)
+        if type(value) is complex:
+            self.raise_syntax_error_known_location("real number required in complex literal", number)
+        return value
 
-    def ensure_imaginary(self, number_str: str):
-        number = ast.literal_eval(number_str)
-        if type(number) is not complex:
-            self.raise_syntax_error("imaginary number required in complex literal")
-        return number
+    def ensure_imaginary(self, number:  ast.Constant):
+        value = ast.literal_eval(number.string)
+        if type(value) is not complex:
+            self.raise_syntax_error_known_location("imaginary number required in complex literal", number)
+        return value
 
     def generate_ast_for_string(self, tokens):
         """Generate AST nodes for strings."""
-        err_msg = ''
-        line = 1
+        err = None
+        line_offset = tokens[0].start[0]
+        line = line_offset
         col_offset = 0
-        source = ''
+        source = "(\\n"
         for t in tokens:
             n_line = t.start[0] - line
             if n_line:
                 col_offset = 0
-            source += """\n""" * n_line + ' ' * (t.start[1] - col_offset) + t.string
+            source += """\\n""" * n_line + ' ' * (t.start[1] - col_offset) + t.string
             line, col_offset = t.end
-        if source[0] == ' ':
-            source = '(' + source[1:]
-        else:
-            source = '(' + source
-        source += ')'
+        source += "\\n)"
         try:
             m = ast.parse(source)
         except SyntaxError as e:
-            err_msg = e.args[0]
-            # Identify the line at which the error occurred to get a more
-            # accurate line number
-            for t in tokens:
-                try:
-                    m = ast.parse(t.string)
-                except SyntaxError:
-                    break
+            err = e
 
         # Avoid getting a triple nesting in the error report that does not
         # bring anything relevant to the traceback.
-        if err_msg:
-            self.raise_syntax_error_known_location(err_msg, t)
-            raise self._exception
-
-        return m.body[0].value
+        if err is not None:
+            args = (err.filename, err.lineno + line_offset - 2, err.offset, err.text)
+            if sys.version_info >= (3, 10):
+                args += (err.end_lineno + line_offset - 2, err.end_offset)
+            raise SyntaxError(err.msg, args)
+
+        node = m.body[0].value
+        ast.increment_lineno(node, line_offset - 2)
```

Comment:
Why do we need to do this?

Reply:
I simplified the code to avoid generating many blank line so this is my way to get back the right lineno.

Reply:
Added a comment